libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/offroute/OffRouteDetectorTest.java

```diff
@@ -1,16 +1,331 @@
 package com.mapbox.services.android.navigation.v5.offroute;
 
-import static junit.framework.Assert.assertNotNull;
+import android.location.Location;
+import android.support.annotation.NonNull;
 
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.mapbox.api.directions.v5.DirectionsAdapterFactory;
+import com.mapbox.api.directions.v5.models.DirectionsResponse;
+import com.mapbox.api.directions.v5.models.DirectionsRoute;
+import com.mapbox.api.directions.v5.models.LegStep;
+import com.mapbox.core.constants.Constants;
+import com.mapbox.geojson.LineString;
+import com.mapbox.geojson.Point;
+import com.mapbox.services.android.navigation.BuildConfig;
 import com.mapbox.services.android.navigation.v5.BaseTest;
+import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigationOptions;
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;
+import com.mapbox.services.android.navigation.v5.utils.RingBuffer;
+import com.mapbox.turf.TurfConstants;
+import com.mapbox.turf.TurfMeasurement;
 
+import org.junit.Before;
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+import java.io.IOException;
+import java.util.List;
 
+import static junit.framework.Assert.assertFalse;
+import static junit.framework.Assert.assertNotNull;
+import static junit.framework.Assert.assertTrue;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(constants = BuildConfig.class, manifest = Config.DEFAULT_MANIFEST_NAME)
 public class OffRouteDetectorTest extends BaseTest {
 
+  private static final String DIRECTIONS_PRECISION_6 = "directions_v5_precision_6.json";
+
+  private OffRoute offRouteDetector;
+  private MapboxNavigationOptions options;
+  private RingBuffer<Integer> distances;
+
+  @Mock
+  Location mockLocation;
+  @Mock
+  RouteProgress mockProgress;
+  @Mock
+  OffRouteCallback mockCallback;
+
+  @Before
+  public void setup() throws Exception {
+    MockitoAnnotations.initMocks(this);
+
+    offRouteDetector = new OffRouteDetector();
+    options = MapboxNavigationOptions.builder().build();
+    distances = new RingBuffer<>(3);
+  }
+
   @Test
   public void sanity() throws Exception {
-    OffRoute offRoute = new OffRouteDetector();
-    assertNotNull(offRoute);
+    assertNotNull(offRouteDetector);
+  }
+
+  @Test
+  public void invalidOffRoute_onFirstLocationUpdate() throws Exception {
+    boolean isUserOffRoute = offRouteDetector.isUserOffRoute(
+      mockLocation, mockProgress, options, distances, mockCallback
+    );
+    assertFalse(isUserOffRoute);
+  }
+
+  @Test
+  public void validOffRoute_onMinimumDistanceBeforeReroutingPassed() throws Exception {
```

Comment:
Generally is preferable having only 1 assertion per test 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=57

Reply:
Why having comments in a test?

Reply:
Normally is good to structure the tests following the triple A 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40 http://wiki.c2.com/?ArrangeActAssert
If the `A`rrange part gets too long we should create factory or builder methods.

Reply:
Yeah some of these assertions need to be checked after multiple location updates given to the detector IMO - in order to properly check the detector is acting correctly after certain update scenarios

Reply:
The problem of having multiple assertions are:
- It's annoying to have to look at the stacktrace to know what's failing (which assertion failed?)
- You don't have information about the remaining assertions (because when an assertion fails it breaks the execution of the test)

That's why is preferable to have only 1 assertion per test and one test per scenario. Basically, if one of them fails you'll know at a glance what in the code might have caused that failure and in which part (class, method, etc.).